guides/create_an_app/4-query.md

```diff
@@ -1,5 +1,88 @@
+## Retrieving from database
+
+Everytime when you refresh your page, you lose the message history and that is not ideal if you want to know what has been discussed. We have all these stored messages in our database, and we want to retrieve it when someone logs in. Therefore, we will need to retrieve these messages and display them in the chat dialog.
+
+## Perform a callback `after_join`.
+Go to the join function, and add the line just below `if authorized?(payload) do`
+
+We are going to send a message contaning our self *PID(Process Identifier)* as the first parameter. The second parameter is a message to run a function called `after_join`. This function is identified by an atom type.
+Add this line to function inside `lib/workshops_app_web/channels/general_channel.ex`
+
+```elixir
+  def join("random:lobby", payload, socket) do
+    if authorized?(payload) do
+      send self(), :after_join #<-- Add this line
+      {:ok, socket}
+    else
+      {:error, %{reason: "unauthorized"}}
+    end
+  end
+```
+
+### What does send do?
+`send` sends a message `:after_join`, with it's PID generated by the function `self()`. We want to be able to run that after join task asynchronously. Lets look at how we should receive the `:after_join` message.
+
+## Receiving a message in a callback
+We now know that a message identified by the atom `:after_join` is being sent. Phoenix provides us with a function called `handle_info` to be able to receive that message and do something with it. Insert the following code after the `join` function.
+
+```elixir
+  def handle_info(:after_join, socket) do
+    messages =
+      WorkshopsApp.Conversations.list_messages()
+      |> Enum.map(fn(m) -> %{message: m.message, name: m.name} end)
+
+    push socket, "messages_history", %{messages: messages}
+    {:noreply, socket}
+  end
+```
+
+The only missing thing is `WorkshopsApp.Conversations.list_messages()` (we should get compilation error when function is called but not definied). To have everything working as expected, we need to add function, which will list all messages stored in our database. Let's open `WorkshopsApp.Conversations` module and put there:
+
+```elixir
+def list_messages do
+  Repo.all(Message)
+end
+
+```
+
+**Coach:** WDYT about saying couple words about compilator?
+
+## Retrieving all messages from our database
+Phew, that looked like a lot is going on in the `handle_info` function. Let's breakdown the code to understand what is going on.
```

Comment:
Phew, that looks like a lot is going on in the 'handle_info' function. Let's break down the code to understand what is happening.